File: backend/nfe/src/model/nfe_duplicata_model.py
from src import db
from src.model.nfe_fatura_model import NfeFaturaModel


class NfeDuplicataModel(db.Model):
    __tablename__ = 'nfe_duplicata'

    id = db.Column(db.Integer, primary_key=True)
    numero = db.Column(db.String(60))
    data_vencimento = db.Column(db.DateTime)
    valor = db.Column(db.Float)
    id_nfe_fatura = db.Column(db.Integer, db.ForeignKey('nfe_fatura.id'))

    nfe_fatura_model = db.relationship('NfeFaturaModel', foreign_keys=[id_nfe_fatura])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_fatura = data.get('idNfeFatura')
        self.numero = data.get('numero')
        self.data_vencimento = data.get('dataVencimento')
        self.valor = data.get('valor')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeFatura': self.id_nfe_fatura,
            'numero': self.numero,
            'dataVencimento': self.data_vencimento.isoformat(),
            'valor': self.valor,
            'nfeFaturaModel': self.nfe_fatura_model.serialize() if self.nfe_fatura_model else None,
        }

File: frontend/nfe/lib/app/data/model/nfe_duplicata_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:nfe/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class NfeDuplicataModel {
	int? id;
	int? idNfeFatura;
	String? numero;
	DateTime? dataVencimento;
	double? valor;
	NfeFaturaModel? nfeFaturaModel;

	NfeDuplicataModel({
		this.id,
		this.idNfeFatura,
		this.numero,
		this.dataVencimento,
		this.valor,
		this.nfeFaturaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero',
		'data_vencimento',
		'valor',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero',
		'Data Vencimento',
		'Valor',
	];

	NfeDuplicataModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeFatura = jsonData['idNfeFatura'];
		numero = jsonData['numero'];
		dataVencimento = jsonData['dataVencimento'] != null ? DateTime.tryParse(jsonData['dataVencimento']) : null;
		valor = jsonData['valor']?.toDouble();
		nfeFaturaModel = jsonData['nfeFaturaModel'] == null ? NfeFaturaModel() : NfeFaturaModel.fromJson(jsonData['nfeFaturaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeFatura'] = idNfeFatura != 0 ? idNfeFatura : null;
		jsonData['numero'] = numero;
		jsonData['dataVencimento'] = dataVencimento != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataVencimento!) : null;
		jsonData['valor'] = valor;
		jsonData['nfeFaturaModel'] = nfeFaturaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeFatura = plutoRow.cells['idNfeFatura']?.value;
		numero = plutoRow.cells['numero']?.value;
		dataVencimento = Util.stringToDate(plutoRow.cells['dataVencimento']?.value);
		valor = plutoRow.cells['valor']?.value?.toDouble();
		nfeFaturaModel = NfeFaturaModel();
		nfeFaturaModel?.numero = plutoRow.cells['nfeFaturaModel']?.value;
	}	

	NfeDuplicataModel clone() {
		return NfeDuplicataModel(
			id: id,
			idNfeFatura: idNfeFatura,
			numero: numero,
			dataVencimento: dataVencimento,
			valor: valor,
		);			
	}

	
}